### PR TITLE
Combine deconstruction assignment and declaration, and support discards (merging to master)

### PR DESCRIPTION
**Pull request by jcouv**
This is merging Neal's recent change in the features/wildcard branch (https://github.com/dotnet/roslyn/pull/15548) back into master.

**Customer scenario**

This parsing fix addresses various ambiguities that would be backcompat issues once we allow mixing of deconstruction declarations and assignments. Those scenarios still report errors for now.
For instance, `(a < b, c > d, e) = ...;`. Are there 2 or 3 elements on the left?

**Bugs this fixes:** 

Fixes https://github.com/dotnet/roslyn/issues/14794
Fixes https://github.com/dotnet/roslyn/issues/14832

**Workarounds, if any**

No workaround. If we let customers type such code in C#7.0, we can't change how it gets interpreted in a later version.

**Risk**

Low. The change is mostly limited to deconstruction code paths.

**Performance impact**

Low. The parsing has actually become more local (less parsing ahead) which should be a minor improvement.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

This is a ripple effect of our decision to rely on declaration expressions, rather than specialized deconstruction syntax.  The declarative expressions change was made due to direct customer feedback about the design of the feature. 

**How was the bug found?**

Know work item (listed in work breakdown https://github.com/dotnet/roslyn/issues/14832)

### CODE REVIEW

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 1]
```diff
@@ -430,14 +430,14 @@ class Variable
                 Await state.AssertSelectedCompletionItem(displayText:="as", isSoftSelected:=True)
 
                 state.SendTypeChars(", va")
-                Await state.AssertSelectedCompletionItem(displayText:="var", isSoftSelected:=True)
+                Await state.AssertSelectedCompletionItem(displayText:="var", isHardSelected:=True)
```
Comment (jcouv) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: FYI @CyrusNajmabadi Merging Neal's latest parsing changes with the completion tests I had added in master requires those tests to be updated. I think the updated expectations are an improvement.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler This is a merge from `features/wildcard` branch back into `master`. I'm not sure what review procedure applies.
The merge was mostly automatic with only a couple small conflicts. After that I had to make a few test adjustments, which I kept in a the third commit.

**Comment by AlekseyTs** [comment]
The fix-up LGTM, it is my understanding the the initial commit has been signed-off on earlier.

**Comment by jaredpar** [comment]
CC @MattGertz for ask mode approval

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb (hunk 1 above)
Comment (jcouv) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: @CyrusNajmabadi Ping

**Pull request merged**